core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -114,34 +113,24 @@ class LogCleaner(initialConfig: CleanerConfig,
 
   /* a metric to track the maximum utilization of any thread's buffer in the last cleaning */
   newGauge("max-buffer-utilization-percent",
-           new Gauge[Int] {
-             def value: Int = cleaners.map(_.lastStats).map(100 * _.bufferUtilization).max.toInt
-           })
+    () => cleaners.iterator.map(100 * _.lastStats.bufferUtilization).max.toInt)
+
   /* a metric to track the recopy rate of each thread's last cleaning */
-  newGauge("cleaner-recopy-percent",
-           new Gauge[Int] {
-             def value: Int = {
-               val stats = cleaners.map(_.lastStats)
-               val recopyRate = stats.map(_.bytesWritten).sum.toDouble / math.max(stats.map(_.bytesRead).sum, 1)
-               (100 * recopyRate).toInt
-             }
-           })
+  newGauge("cleaner-recopy-percent", () => {
+    val stats = cleaners.map(_.lastStats)
```

Comment:
This will still create an intermediate collection, no? Why not use `cleaners.iterator.map`, turn stats into an iterator... and avoid creating the extra collection completely?

Reply:
Yes, good catch. I missed that in the initial change.. Will update.

Reply:
I went to make this change and then I remembered why I didn't do it in the first place. The iterator can only be used once and we need two iterations. We would need to allocate two separator iterators. But `cleaners` should be a small collection (size 1 by default) so the additional code complexity didn't seem worth it.